File: Java/Sorting/ShellSort.java
/*
Approach:
- Shell Sort is an extension of the insertion sort algorithm that works by sorting subarrays.
- It starts by sorting elements that are far apart, and gradually reduces the gap between elements to be compared.
- The algorithm uses a sequence of gaps or steps, which depends on the chosen gap sequence.

Time Complexity: O(n^2) in the worst case. With the right gap sequence, it can be improved to O(n log^2 n).

Space Complexity: O(1) - It is an in-place sorting algorithm.

Sample Input:
[12, 34, 54, 2, 3]

Sample Output:
[2, 3, 12, 34, 54]
*/

public class ShellSort {
    public static void shellSort(int[] arr) {
        int n = arr.length;

        // Choose the gap sequence. Here, we use the Knuth sequence.
        int gap = 1;
        while (gap < n / 3) {
            gap = 3 * gap + 1;
        }

        while (gap > 0) {
            for (int i = gap; i < n; i++) {
                int temp = arr[i];
                int j = i;
                while (j >= gap && arr[j - gap] > temp) {
                    arr[j] = arr[j - gap];
                    j -= gap;
                }
                arr[j] = temp;
            }
            gap /= 3;
        }
    }

    public static void main(String[] args) {
        int[] arr = {12, 34, 54, 2, 3};
        shellSort(arr);

        System.out.print("Sorted Array: [");
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i]);
            if (i < arr.length - 1) {
                System.out.print(", ");
            }
        }
        System.out.println("]");
    }
}

File: C++/Sorting/ShellSort.cpp
//Shell Sort is an in-place comparison sort algorithm that improves upon the insertion sort by comparing elements that are far apart and then progressively reducing the gap between elements to be compared. Here's the algorithm, C++ code, and a time complexity analysis for Shell Sort:

//Algorithm:

//Choose a gap sequence (commonly used is the Knuth sequence or Sedgewick sequence).
//Starting with the first gap, compare and swap elements that are the gap distance apart.
//Continue this process for all elements in the array.
//Reduce the gap by a factor (commonly 2), and repeat the process until the gap becomes 1.
//Finally, apply a standard insertion sort with a gap of 1 to ensure a fully sorted array.
//C++ Code
#include <iostream>
using namespace std;

void shellSort(int arr[], int n) 
{
     //Start with a large gap and reduce it over time
    for (int gap = n / 2; gap > 0; gap /= 2) {
        // Perform an insertion sort for the elements at each gap
        for (int i = gap; i < n; i++) {
            int temp = arr[i];
            int j;
            for (j = i; j >= gap && arr[j - gap] > temp; j -= gap) {
                arr[j] = arr[j - gap];
            }
            arr[j] = temp;
        }
    }
}

int main() {
    int arr[] = {12, 34, 54, 2, 3};
    int n = sizeof(arr) / sizeof(arr[0]);

    cout << "Original array: ";
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }

    shellSort(arr, n);

    cout << "\nSorted array: ";
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }

    return 0;
}
//Time Complexity Analysis:

//The time complexity of Shell Sort depends on the choice of gap sequence. Using the Sedgewick sequence or Knuth sequence generally results in better performance. The worst-case time complexity for the Shell Sort is roughly O(n^2), which occurs when the gap sequence is poorly chosen.

//In the best case, with the right gap sequence, it can be close to O(n log^2 n).
//In the average case, it's usually better than the worst case but not as good as the best case.
//The space complexity of Shell Sort is O(1) since it sorts the elements in-place.
//Overall, Shell Sort is an efficient sorting algorithm and can be useful when you want a simple sorting method with better performance than insertion sort.
